Adding seamless Windows VM (with docker) launcher as native app

sspaeti opened the pull request:
This is my implementation of a starting a Windows VM with one launcher command. Not sure if that is too heavy of a feature for Omarchy, but I thought I leave it here in case anybody would like or have the requirement to use an application that only runs on Windows. This way, this way we can quickly ramp up a VM with docker and connects via RDP, and automatically close when we close the TUI.

## What's Added
- **`omarchy-launch-ms-windows`** - Smart launcher script that manages Windows docker container and connects via rdesktop
- **Windows VM desktop entry** - One-click access from application launcher
- **Automatic container management** - Handles startup, health checks, and cleanup
- **Persistent storage** - VM data stored in `~/.windows` directory

## Features
- **One-click launch** from application menu
- **Health monitoring** - Waits for container and RDP to be fully ready
- **Auto-cleanup** - VM shuts down when RDP session ends
- **Smart container handling** - Removes existing containers and starts fresh
- **Integration with UWSM** - Properly managed as Hyprland application

## Requirements
- `rdesktop` - For RDP client connections
- `netcat` (`nc`) - For port availability checking
- `curl` - For web interface health checks
- Docker (already provided by Omarchy)

## Technical Details
Uses the excellent [dockur/windows](https://github.com/dockur/windows) Docker image which handles Windows 11 ISO download, installation, and configuration automatically. First-time setup takes 30-60
   minutes for Windows installation, subsequent launches are much faster (a couple of seconds).

**Demo**: See it in action at https://x.com/sspaeti/status/1961826772682490032
  
  
## Open
Things not finalized yet where I'd need help:
- there's no migration script yet as I didn't know if this featues is wanted or not
- not sure if the relative path of icons/windows.png would work.

rmacklin commented:
> Not sure if that is too heavy of a feature for Omarchy

Probably depends if it runs Lightroom well

bjarneo commented:
This idea is interesting. I really like it as some software can't run on Linux, yet.

It could be done running a docker instance of osx as well: https://github.com/sickcodes/Docker-OSX

sspaeti commented:
> > Not sure if that is too heavy of a feature for Omarchy
> 
> Probably depends if it runs Lightroom well

I don't have an Adobe license, but I'd imagine it works. I'm running MS Word when editing some large word documents. Works like a charm, even the CPU/RAM usage is very consistent.

> It could be done running a docker instance of osx as well: [sickcodes/Docker-OSX](https://github.com/sickcodes/Docker-OSX)

I saw that, indeed. I don't have that requirement, but we could set it up as well if anyone has macOS software, making the initial transition easier.

bjarneo commented:
I did not add all the context I had in mind. I don't think osx needs to be
part of this PR. Was more of a, cool, maybe something similar can be done
for osx separately for those in need of that.

On Sun, 31 Aug 2025, 13:01 Simon Späti, ***@***.***> wrote:

> *sspaeti* left a comment (basecamp/omarchy#1333)
> <https://github.com/basecamp/omarchy/pull/1333#issuecomment-3240050920>
>
> Not sure if that is too heavy of a feature for Omarchy
>
> Probably depends if it runs Lightroom well
>
> I don't have an Adobe license, but I'd imagine it works. I'm running MS
> Word when editing some large word documents. Works like a charm, even the
> CPU/RAM usage is very consistent.
>
> It could be done running a docker instance of osx as well:
> sickcodes/Docker-OSX <https://github.com/sickcodes/Docker-OSX>
>
> I saw that, indeed. I don't have that requirement, but we could set it up
> as well if anyone has macOS software, making the initial transition easier.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/basecamp/omarchy/pull/1333#issuecomment-3240050920>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAK2PTW45FEDGV5FVUXDZPD3QLIYDAVCNFSM6AAAAACFG56HWKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENBQGA2TAOJSGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

sspaeti commented:
Just for reference, there's another project that uses docker https://github.com/winapps-org/winapps, and they say Lightroom is supported -> so I guess it will work fine @rmacklin and @dhh :)

Thanks to [k3nundrum](https://discord.com/channels/1390012484194275541/1410499858170646610/1411769697946439741) from the Discord community.

geiman commented:
I was actually working on testing WinApps today on Omarchy.  Haven't tested Lightroom yet, but gave Photoshop a try.  Wasn't really usable in it's current form in my opinion.  Menus and other UI elements show up blurry 50-75% of the time for some reason, and manipulating elements in a project was very difficult.  Guessing it's something with how Hyprland renders the window.  Using this bind to disable the blur doesn't seem to do anything to help:

https://github.com/basecamp/omarchy/issues/1295

I'm guessing Lightroom will be the same.  

But O365 apps work great - Word, Excel, PowerPoint and Outlook all functioned fine.  Haven't tested Teams yet, but I imagine it will be fine.

sspaeti commented:
Update: I changed the setup to use a proper `docker-compose.yml` as this seem to work better (https://github.com/basecamp/omarchy/pull/1333#issuecomment-3247247356, I also use this version), plus it's easier to add parameters such as CPU or MEMORY or the windows KEY.

One thing I'm unsure - we have two different modes that you also ran into it:
1. initial install. It takes a while until it downloads everything and installs all. This can be done with `docker compose up`.
2. once everything is installs initially, the added desktop app `applications/windows-vm.desktop` in combination with the `bin/omarchy-launch-ms-windows` will handle the automatically starting docker compose and once ready connect via Rdesktop. See demo here once it is installed: https://x.com/sspaeti/status/1961826772682490032

But should we have a `bin/omarchy-install-ms-windows` script that does the installation? In theory, it will also install the docker container when you run the desktop app, but it's not that transparent, and if something goes wrong, you need to check http://127.0.0.1:8006/. So not sure what the best "workflow" for that is.

Happy to get feedback. But I think it's not far away to be super useful to people.

dhh commented on config/windows-vm/docker-compose.yml:
```diff
@@ -0,0 +1,22 @@
+      VERSION: "11"
+      # RAM_SIZE: "12G"
+      # CPU_CORES: "4"
+      # KEY: "XXXXX-XXXXX-XXXXX-XXXXX-XXXXX"  # Uncomment and add your Windows product key for initial installation
```
Ah, that's a good idea. Yeah, I think we need to prompt for the key, and then folks can do whatever they want from there.

dhh commented on install/omarchy-other.packages:
```diff
@@ -30,13 +31,14 @@ macbook12-spi-driver-dkms
 pipewire-pulse
 qt5-remoteobjects
 qt6-wayland
+rdesktop
```
We should install rdesktop on demand the first time this is run.

sspaeti commented:
OK, I added the installation within the bash script `bin/omarchy-launch-ms-windows` and added the wrong KEY to prompt during installation, which looks like this:

<img width="1270" height="871" alt="image" src="https://github.com/user-attachments/assets/9955dc7d-97f5-44c9-9048-bfacc0fbbb12" />

You can then fill in your key (or a dev key as the https://github.com/dockur/windows is using as default:  https://gist.github.com/rvrsh3ll/0810c6ed60e44cf7932e4fbae25880df#kms-client-product-keys)

dhh commented:
That's great. Feel much better about that!

In terms of setup, I'd do Install > Windows to setup the docker stuff and then make the docker stuff persistent across reboots by default. And then we add Windows WM as a webapp against localhost:8006 in that install setup too. Need to add Remove > Windows too then.

dhh commented:
We should copy the .desktop file to `~/.local/share/applications`. That's where the other Omarchy ones are.

I think we can just go with "Windows" instead of "Windows WM". I like it under install 👌

We def need to test that the out-of-the-box is totally smooth.

This seems super promising!

sspaeti commented:
OK done. And yes, please, I'd be happy if someone could test it sorrowly and give feedback.

ryanrhughes commented:
@dhh in thinking about this; couple of conceptual considerations we should probably weigh here:

- We copy all .desktop files in `install/config/mimetypes.sh` which would bring this in without having installed it
- The `omarchy-launch-windows-vm` would also be a bin available without actually functioning and need error handling. 

We could either have it available right from the start; in which case `omarchy-launch-windows-vm` should have a gate to make sure that we can reasonably assume the install has been ran and will have a successful launch, otherwise throw an option to run the install.

This route also kind of removes the need for an option in the install menu since it'd be kind of an implicit path on first execution.

If we don't want to show it at all unless people install it, I think I'd probably suggest we remove the .desktop file and have the installer write the file using `tee` so it only exists after install. The icon is fine and doesn't cause any real issue.

sspaeti commented:
> When you start Windows either using `omarchy-launch-windows-vm` or from Walker, it opens a full-screen window after a few seconds, and I think this behavior could happen only when launching from Walker.

> Lastly, is it possible to make the window that opens automatically scale the resolution dynamically? So when not in full-screen it does not clip the bottom of the Windows sceen:

The fullscreen is `+f`, which can be removed when connecting. Also, the resolution is hard-coded at the moment with this command: 
```sh
RDP_COMMAND="xfreerdp3 /u:docker /p:admin /v:127.0.0.1:3389 /size:1920x1200 /scale:140" +f -grab-keyboard /cert:ignore"
```

This can be customized, e.g. I changed mine to `/size:1600x1000 /scale:140"`. This could be a variable, or read from `hypr monitors`, but I'm afraid that it might get **too complex** and not always work. Especially what if you have two or more screens, which one do we choose? Might be much easier to just customize it once in your configs, and then it will work (?) 🤔

> So I could simply access localhost:8006 to use it.
The web access on port 8006 is only for checking or troubleshoot, for using, you use RDC on port `3389`, much better quality.

> Another reason for that is when I close Windows the container automatically stops (which is expected), but every time I need to use it again (which happens sporadically throughout the day) I need to wait for the container and windows to load again

Fair point. To avoid this, we could also just keep the container running and have another command to turn on/off Windows VM, and the launch would just connect the RDC? But I don't like on my machine to have a docker container running, that uses lots of RAM/CPU, better to quickly restart.  

> there's no feedback that anything is happening. Maybe a simple dialog could appear saying "Windows container starting, please wait"

Good point. I think once we merge it, we could always update and work on smaller details. To me, most important that it works reliable and people can use Windows as simple and easy as possible.

3bit commented:
Have you considered using WinBoat for this feature?
https://www.winboat.app/
https://github.com/TibixDev/winboat

Winboat adds many benefits over a simple (Docker) Windows VM, like running single Windows applications (via RDP), filesystem integration, automated installs, smartcard and USB passthrough and more.
